### PR TITLE
Split wasmstation into separate crates for web support.

### PR DESCRIPTION
**Pull request by grantshandy**
Wasmstation in WebAssembly isn't really very difficult to create at this point, if this PR is merged I think I can complete it this weekend. I tried to keep this PR small, but really there's not much I can do when it moves most source files in the repo around and renames packages. All tests pass and it is just as successful with wasm4-test-cart.

**Here are my intentions for the crates created in this PR:**
 - `crate wasmstation-core` - The core WASM-4 types, functions, and constants in all `wasmstation` programs. Extremely fast and robust: polished, no panics, no platform-specific code, minimal/no heap allocations.
 - `crate wasmstation-desktop` - Desktop-specific renderer (e.g. `launch`). In the future this can have simple flags like `sdl2` and `wgpu`.
 - `crate wasmstation-wasmer` - The default `Backend` for `wasmstation-desktop` (maybe it should be re-exported by `wasmstation-desktop`?).

*and soon:*
 - `crate wasmstation-web` - Includes `fn launch_on_canvas(backend: impl Backend, id: &str)` and `struct WebBackend`.

**Changes:**
 - `wasmstation::renderer` has been moved to the `wasmstation-desktop` crate. SDL2 is the most platform-specific dependency and it is only used here, so I'd like to separate it from the other pieces of code.
 - `wasmstation::backend::wasmer` has been moved to the `wasmstation-wasmer` crate, mainly because it requires heavy usage of `libstd`.
 - `wasmstation` has been renamed to `wasmstation_core`.
 - `wasmstation::console` has been moved to the crate root of `wasmstation_core`.
 - Removed `#![allow(unused_variables, dead_code)]` - I understand why you might want to add this at the beginning of a project when you're adding incomplete features, but this allow rule kept in a *lot* of dead and unused code. Fixing these warnings made me do a few things:
   - Tests in individual `wasmstation::console::framebuffer::foo::tests` modules and related data types have been moved to `wasmstation_core::framebuffer::tests`.
     - This was done because about half the things in `wasmstation::console::framebuffer` were solely for tests (e.g. `ArrayScreen`, `as_fb_vec` variants).
     - This also improves general readability quite a bit 😄.
   - Some struct fields and variables were changed to `_foo` instead of `foo` because they weren't used anywhere in the crate (and they weren't related to any public interfaces). I want to remove these types of thing in a future PR but I've silenced the error here so I can get as much done in as little controversy.
 - Workspace default values. This enables us to only use a single `bytemuck`/`log`/`anyhow` version and edition/license/repository metadata across the workspace for compatibility.
 - Add `MemoryViewSource` in `wasmstation_wasmer` because `Source` is no longer implemented in the same crate as `WasmerBackend`. This is something that I probably should have done in preparation for this during #42.
 - Fix some basic "uncontroversial" `clippy` lints such as:
   - for loop => enumerated iter in `Source::items_at`
   - `mem::swap` instead of "variable musical chairs" in `line`.
   - Unnecessary `&`s and explicit `return`s.
   - `impl Default` => `#[derive(Default)]`
   - etc. etc. etc.
 - Standardize some documentation.

### CODE REVIEW

**Review by upachler — COMMENTED** [review]
Yes, this is huge, so I'm getting the feeling I cannot really be a thorough as I should be. However, I'm sure this all works like you said, and in general I'm with you on the goals you want to achive with this.

However I do think this is also an opportunity to remove dead code; I've pointed out a couple of things in this regard (e.g. the attempt to have a more byte that bit oriented blit function, which did not work and is now commented out anyways, or unused variables that the compiler would optimize out anyways; may well be removed entirely).

File: core/src/framebuffer/line.rs [hunk 1]
```diff
@@ -72,7 +67,7 @@ fn line_impl<T: Screen>(
     }
 }
 
-/// Draw a horizontal line between `(x, y)` and `(x + len - 1, y)`.
```
[the remark this answers is not in the document]
Comment (grantshandy) [reply core/src/framebuffer/line.rs:75]: Resolved in [45f6fdc](https://github.com/wasmstation/wasmstation/pull/44/commits/45f6fdcb295967adf66828a7fa701b782424bddd) 👍🏽.

File: core/src/framebuffer/line.rs [hunk 2]
```diff
@@ -1,13 +1,13 @@
 use crate::{wasm4::FRAMEBUFFER_SIZE, Sink, Source};
+use core::mem;
 
 use super::{
     remap_draw_color, set_pixel_impl, set_pixel_unclipped_impl, Screen, Wasm4Screen, DRAW_COLOR_1,
 };
 
 /// Draw a line between two points.
-///
```
[the remark this answers is not in the document]
Comment (grantshandy) [reply core/src/framebuffer/line.rs:8]: Resolved in [45f6fdc](https://github.com/wasmstation/wasmstation/pull/44/commits/45f6fdcb295967adf66828a7fa701b782424bddd) 👍🏽.

File: core/src/audio.rs [hunk 3]
```diff
@@ -279,8 +268,7 @@ struct AudioChannelState {
 #[derive(Default)]
 struct AudioChannel {
     state: AudioChannelState,
-
-    start_frame: FrameCount,
+    _start_frame: FrameCount,
```
[the remark this answers is not in the document]
Comment (grantshandy) [reply core/src/audio.rs]: Resolved in [45f6fdc](https://github.com/wasmstation/wasmstation/pull/44/commits/45f6fdcb295967adf66828a7fa701b782424bddd) 👍🏽.

File: core/src/audio.rs [hunk 4]
```diff
@@ -321,7 +309,7 @@ impl AudioChannel {
         }
 
         // render sample
-        let gen = &mut self.generator;
```
[the remark this answers is not in the document]
Comment (grantshandy) [reply core/src/audio.rs:324]: Resolved in [45f6fdc](https://github.com/wasmstation/wasmstation/pull/44/commits/45f6fdcb295967adf66828a7fa701b782424bddd) 👍🏽.

File: core/src/framebuffer/blit.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+            let mut byte = sprite.item_at((pixel_index >> 2) as usize).unwrap();
+            byte >>= 6 - ((pixel_index & 0x03) << 1);
+            byte & 0x03
+        }
+        #[cfg(test)]
+        PixelFormat::Framebuffer => panic!("invalid pixel format for reading sprite data"),
+    }
+}
+
+// #[allow(clippy::too_many_arguments)]
```
[the remark this answers is not in the document]
Comment (grantshandy) [reply core/src/framebuffer/blit.rs]: Resolved in [45f6fdc](https://github.com/wasmstation/wasmstation/pull/44/commits/45f6fdcb295967adf66828a7fa701b782424bddd) 👍🏽.

**Comment by grantshandy** [comment]
Sounds good, thanks for pointing out some dead code 👍🏽. I didn't want to remove some of those values in fear that they would inadvertently break something that I didn't write.

**Pull request merged**